### ISSUE TITLE
Support multiple instances of the same package/provider

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi#1221]
Currently, we do not support deploying resources to two different AWS accounts, two Kubernetes clusters, etc. from the same Pulumi program.  This is a requirement for many practical applications of Pulumi.

To support this, we'll need to have a way to create additional instances of a package/provider - such as:

```
import * as aws from "@pulumi/aws"

let usEast1Aws = aws.New({ ... config options ... });
let usWest1Aws = aws.New({  ... config options ...});
let instance = new usEast1Aws.ec2.Instance(...);
```

This poses a few questions:
1. How does this interact with the config system?  Is ambient config always applied only to the "default" instance?
2. How do these separate instances manage their identity across the language<->engine<->provider boundary?  Specifically, how do RPC calls track which instance of the resource provider they should be referring to?
2. Is this a convention-based model that each package or provider can choose how to expose to users, or something that is formalized by `@pulumi/pulumi`?
3. How is TypeScript typing handled for these dynamically created instances of the module?

**Comment by Jtango18** [comment]
I've run up against this today. 

I'm deploying all my code in to the Sydney Region (AP-SOUTHEAST-2), but part of my infrastructure is a CloudFront connected Lambda@Edge function. These sorts of functions MUST be deployed in to the US East (N. Virginia) Region (US-EAST-1)

Happy to be a sounding board if you need user feedback on this.

**Comment by pgavlin** [comment]
Fixed with #1695.
